lib/client/conntest/ssh.go

```diff
@@ -49,30 +63,243 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
+	if err := s.userClt.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key, err := client.GenerateRSAKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	currentUser, err := s.userClt.GetCurrentUser(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	certs, err := s.userClt.GenerateUserCerts(ctx, proto.UserCertsRequest{
+		PublicKey:              key.MarshalSSHPublicKey(),
+		Username:               currentUser.GetName(),
+		Expires:                time.Now().Add(time.Minute).UTC(),
+		ConnectionDiagnosticID: connectionDiagnosticID,
+	})
+	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	// TODO(marco): test ssh connection
-	// - create certificate for user with extra field
-	// - create ssh client using that certificate
-	// - if connection fails because of a network error, a trace must be included indicating that the host is not reachable
-	// - other traces will be added by the Node itself (rbac checks, principal)
-	connectionDiagnostic.SetMessage("dry-run")
-	connectionDiagnostic.SetSuccess(true)
+	key.Cert = certs.SSH
+	key.TLSCert = certs.TLS
+
+	certAuths, err := s.userClt.GetCertAuthorities(ctx, types.HostCA, false /* loadKeys */)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	hostkeyCallback, err := hostKeyCallbackFromCAs(certAuths)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.TrustedCA = auth.AuthoritiesToTrustedCerts(certAuths)
+
+	keyAuthMethod, err := key.AsAuthMethod()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clusterName, err := s.userClt.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clientConfTLS, err := key.TeleportClientTLSConfig(nil, []string{clusterName.GetClusterName()})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	host := req.ResourceName
+	hostPort := 0
+	tlsRoutingEnabled := req.TLSRoutingEnabled
+
+	parsedProxyHostAddr, err := client.ParseProxyHost(req.ProxyHostPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	webProxyAddr := parsedProxyHostAddr.WebProxyAddr
+	sshProxyAddr := parsedProxyHostAddr.SSHProxyAddr
+
+	key.KeyIndex = client.KeyIndex{
+		Username:    req.SSHPrincipal,
+		ProxyHost:   webProxyAddr,
+		ClusterName: clusterName.GetClusterName(),
+	}
+
+	if !tlsRoutingEnabled {
+		node, err := s.proxyClt.GetNode(ctx, defaults.Namespace, req.ResourceName)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				connDiag, err := s.userClt.AppendDiagnosticTrace(ctx, connectionDiagnosticID, types.NewTraceDiagnosticConnection(
+					types.ConnectionDiagnosticTrace_RBAC_NODE,
+					"Node not found. Ensure the Node exists and your role allows you to access it.",
+					err,
+				))
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
 
-	if err := s.cd.UpdateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+				return connDiag, nil
+			}
+			return nil, trace.Wrap(err)
+		}
+
+		addrParts := strings.Split(node.GetAddr(), ":")
```

Comment:
Can we reuse SplitHostPort https://github.com/gravitational/teleport/blob/master/lib/utils/utils.go#L304